cmd/etcd-launcher/main.go

```diff
@@ -352,20 +356,46 @@ func (e *etcdCluster) updatePeerURLs(log *zap.SugaredLogger) error {
 	return nil
 }
 
-func initialMemberList(n int, namespace string, useTLSPeer bool) []string {
-	format := "etcd-%d=http://etcd-%d.etcd.%s.svc.cluster.local:2380"
-
-	if useTLSPeer {
-		format = "etcd-%d=https://etcd-%d.etcd.%s.svc.cluster.local:2381"
-	}
-
+func initialMemberList(n int, namespace string, useTLSPeer bool, client ctrlruntimeclient.Client, log *zap.SugaredLogger) []string {
 	members := []string{}
 	for i := 0; i < n; i++ {
-		members = append(members, fmt.Sprintf(format, i, i, namespace))
+		var pod corev1.Pod
+		if err := client.Get(context.Background(), types.NamespacedName{Name: fmt.Sprintf("etcd-%d", i), Namespace: namespace}, &pod); err != nil {
+			log.Warnw("failed to get Pod information for etcd, guessing peer URLs", zap.Error(err))
+			if useTLSPeer {
+				members = append(members, fmt.Sprintf("etcd-%d=https://etcd-%d.etcd.%s.svc.cluster.local:2381", i, i, namespace))
+			} else {
+				members = append(members, fmt.Sprintf("etcd-%d=http://etcd-%d.etcd.%s.svc.cluster.local:2380", i, i, namespace))
+			}
+		} else {
+			// use information on the pod to determine if the plaintext and TLS peer ports are going to be open
+
+			if !hasStrictTLS(&pod) {
+				members = append(members, fmt.Sprintf("etcd-%d=http://etcd-%d.etcd.%s.svc.cluster.local:2380", i, i, namespace))
+			}
+
+			if _, ok := pod.ObjectMeta.Annotations[resources.EtcdTLSEnabledAnnotation]; ok {
+				members = append(
+					members,
+					fmt.Sprintf("etcd-%d=https://etcd-%d.etcd.%s.svc.cluster.local:2381", i, i, namespace),
+				)
+			}
```

Comment:
they shouldn't, or at least they were deliberately not mutually exclusive when I implemented them. The way you handle multiple peer URLs for a single etcd member is having several entries in the `ETCD_INITIAL_CLUSTER` list (I found this behaviour to be somewhat poorly documented and cannot find a reference to this in the docs right now). So, if `etcd-0` has two peer URLs, it would look like this:

```bash
ETCD_INITIAL_CLUSTER=etcd-0=http://etcd-0:2380,etcd-0=https://etcd-0:2381,etcd-1=http://etcd-1:2380,etcd-2=http://etcd-2:2380
```

While the migration is happening, both peer URLs need to be in the cluster list, and as such both conditions will be (and need to be) true.

Reply:
The answer to "why did this even work before" is that `ETCD_INITIAL_CLUSTER` is only used when an etcd node tries to join a cluster, which usually is not the case during the peer TLS migration. So the incorrect list was ignored, unless a node happened to be removed mid-migration.

Reply:
@kron4eg I'm still struggling to find a reference for this, which is crazy to be honest, but here's a few bits from the etcd source code:

- The initial cluster is treated as an `URLsMap`: https://github.com/etcd-io/etcd/blob/bfb9aa42059d84ea60a4cddd75d7f3e94b27935c/server/embed/config.go#L803
- The code doc for `NewURLsMap` even shows an example where two URLs are added for the same name: https://github.com/etcd-io/etcd/blob/bfb9aa42059d84ea60a4cddd75d7f3e94b27935c/client/pkg/types/urlsmap.go#L26-L41 (`mach0=http://1.1.1.1:2380,mach0=http://2.2.2.2::2380,mach1=http://3.3.3.3:2380,mach2=http://4.4.4.4:2380`)

There's actually no other way to pass multiple peer URLs for one etcd member in the `ETCD_INITIAL_CLUSTER` env variable.